### PR TITLE
[c++] Add c++ support for Array data type

### LINKED ISSUE #468: [c++] Add Array data support for c++ binding
### Search before asking

- [x] I searched in the [issues](https://github.com/apache/fluss-rust/issues) and found nothing similar.

### Description

Follow-up to #433  which added end-to-end ARRAY column support in Rust core.

The Python bindings need to be updated to expose array support to users; the C++ bindings have partial passthrough and need full user-facing support.

### Willingness to contribute

- [x] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by charlesdong1991**
### Purpose

Linked issue: close #468 

### Brief change log
- Adds ARRAY data type support to the C++ bindings and cover full write and read paths (by creating ArrayWriter and ArrayView, also expose in public api)
- Support nested arrays on both write and read 
- Extend GenericRow, RowView and LookupResult with array-typed getters and setters

### Tests

All tests are passed

### CODE REVIEW

**Comment by charlesdong1991** [comment]
@fresh-borzoni @leekeiabstraction appreciate it a lot if you can take a look! 🙏 thanks!

File: bindings/cpp/src/table.cpp [hunk 1]
```diff
@@ -85,6 +85,219 @@ int Date::Day() const {
...
+    inner_->aw_set_decimal_str(idx, value);
+}
+
+void ArrayWriter::SetArray(size_t idx, ArrayWriter&& nested) {
+    CHECK_AW("ArrayWriter");
+    if (!nested.inner_) {
+        throw std::logic_error("ArrayWriter::SetArray: nested writer not available");
+    }
+    inner_->aw_set_array(idx, *nested.inner_);
+}
```
Comment (leekeiabstraction) [inline bindings/cpp/src/table.cpp:191]: Not familiar with C++, but it seems like the convention of `left::method(right&& arg)` is for left to take ownership.

https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#rf-consume

Should we call `Destroy()` to ensure no memory leak?

File: bindings/cpp/src/table.cpp [hunk 2]
```diff
@@ -191,6 +404,14 @@ void GenericRow::SetDecimal(size_t idx, const std::string& value) {
     inner_->gr_set_decimal_str(idx, value);
 }
 
+void GenericRow::SetArray(size_t idx, ArrayWriter&& writer) {
```
Comment (leekeiabstraction) [inline bindings/cpp/src/table.cpp:415]: Similarly here on calling `Destroy()`

**Review by leekeiabstraction — COMMENTED** [review]
TY for the PR. Left early comments.

File: bindings/cpp/src/table.cpp (hunk 1 above)
Comment (charlesdong1991) [reply bindings/cpp/src/table.cpp:191]: i dont think it will cause memory leak since when nested goes out of scope at caller, ArrayWriter will run and call destroy, and will free `Box` from rust, so cleanup will happen.

But with another thought, i think it is still worth adding here explicitly, since it is aligned with c++ convention, let me update

File: bindings/cpp/src/table.cpp (hunk 2 above)
Comment (charlesdong1991) [reply bindings/cpp/src/table.cpp:415]: the same as above

File: bindings/cpp/include/fluss.hpp [hunk 3]
```diff
@@ -305,15 +308,29 @@ class DataType {
     }
     static DataType Char(int32_t length) { return DataType(TypeId::Char, length, 0); }
     static DataType Binary(int32_t length) { return DataType(TypeId::Binary, length, 0); }
+    /// Constructs an `ARRAY<element>` type. The element DataType (possibly
+    /// itself an array) is deep-copied into a shared owning handle so that
+    /// copies of the outer DataType remain cheap while the element lives
+    /// as long as any reference exists.
+    static DataType Array(DataType element) {
```
Comment (fresh-borzoni) [inline bindings/cpp/include/fluss.hpp:315]: Should we add element nullability?

Rust's ArrayType::with_nullable(false, e) carries it, by dropping here we silently lose ARRAY<X NOT NULL> on schema round-trip. Probably the same applies to python, do you mind to check?

File: bindings/cpp/include/fluss.hpp [hunk 4]
```diff
@@ -305,15 +308,29 @@ class DataType {
...
+    /// Returns the element type of an ARRAY. Returns `nullptr` for non-array
+    /// types. The returned pointer is valid as long as this DataType (or a
+    /// copy holding the same shared element) is alive.
+    const DataType* element_type() const { return element_type_.get(); }
 
    private:
     TypeId id_;
     int32_t precision_{0};
     int32_t scale_{0};
+    std::shared_ptr<DataType> element_type_;
```
Comment (fresh-borzoni) [inline bindings/cpp/include/fluss.hpp:333]: also is_nullable field needed here

File: bindings/cpp/include/fluss.hpp [hunk 5]
```diff
@@ -535,6 +603,90 @@ struct ScanData {
...
+class ArrayView {
+   public:
+    ~ArrayView() noexcept;
+
+    ArrayView(const ArrayView&) = delete;
+    ArrayView& operator=(const ArrayView&) = delete;
+    ArrayView(ArrayView&& other) noexcept;
+    ArrayView& operator=(ArrayView&& other) noexcept;
+
+    size_t Size() const;
```
Comment (fresh-borzoni) [inline bindings/cpp/include/fluss.hpp]: noexcept for Size and ElementType?

File: bindings/cpp/src/ffi_converter.hpp [hunk 6]
```diff
@@ -20,13 +20,72 @@
...
+/// scalar. A non-array input produces a zero-initialised value (nesting == 0).
+///
+/// Using a flat representation — rather than serialising a recursive
+/// `DataType` — keeps the cxx bridge contract small (four `i32`s inside
+/// `FfiColumn`) while preserving full schema fidelity across the FFI boundary
+/// when paired with rebuild_array_type().
+struct FlattenedArrayType {
+    int32_t nesting{0};
+    int32_t leaf_type{0};
+    int32_t leaf_precision{0};
```
Comment (fresh-borzoni) [inline bindings/cpp/src/ffi_converter.hpp:44]: also thread nullability here

File: bindings/cpp/include/fluss.hpp [hunk 7]
```diff
@@ -535,6 +603,90 @@ struct ScanData {
...
+};
+
+namespace detail {
+template <typename Derived>
+inline ArrayView NamedGetters<Derived>::GetArrayView(const std::string& n) const {
+    return Self().GetArrayView(Self().Resolve(n));
+}
+}  // namespace detail
+
+class ArrayWriter {
```
Comment (fresh-borzoni) [inline bindings/cpp/include/fluss.hpp:656]: we have num elements as well if I recall correctly

File: bindings/cpp/test/test_log_table.cpp [hunk 8]
```diff
@@ -829,3 +829,436 @@ TEST_F(LogTableTest, PartitionedTableAppendScan) {
 
     ASSERT_OK(adm.DropTable(table_path, false));
 }
+
+// ============================================================================
+// Array data type tests
```
Comment (fresh-borzoni) [inline bindings/cpp/test/test_log_table.cpp:836]: I am thinking about these edge cases as well:

- Long string in array (≥8 bytes for variable-length encoding path)                                                                                                                                                                              
- ARRAY<DECIMAL(20+, 5)> for non-compact decimal                                                                                                                                                                                                 
- ARRAY<TIMESTAMP(9)> for non-compact timestamp                                                                                                                                                                                                  
- Float NaN for  ±Infinity round-trip                                                                                                                                                                                                             
- ARRAY<BINARY(N)> for fixed-length                                                                                                                                                                                        
- SetInt32(1000) on TINYINT must throw

**Review by fresh-borzoni — COMMENTED** [review]
@charlesdong1991 Ty, looks good in general, added minor comments, PTAL

File: bindings/cpp/include/fluss.hpp (hunk 3 above)
Comment (charlesdong1991) [reply bindings/cpp/include/fluss.hpp:315]: Good observation, with another look, it seems nullability is actually not in place for every data type, not limited to Array. The same with python.

I suggest create a separate issue to fix nullability instead, and focus on supporting array type in this PR. 

WDYT? If you agree, i can create 2 issues for python and c++ @fresh-borzoni

File: bindings/cpp/include/fluss.hpp (hunk 4 above)
Comment (charlesdong1991) [reply bindings/cpp/include/fluss.hpp:333]: same above

File: bindings/cpp/include/fluss.hpp (hunk 7 above)
Comment (charlesdong1991) [reply bindings/cpp/include/fluss.hpp:656]: done, i also expose Size to arraywriter for symmetry with arrayview

File: bindings/cpp/test/test_log_table.cpp (hunk 8 above)
Comment (charlesdong1991) [reply bindings/cpp/test/test_log_table.cpp:836]: nice!

File: bindings/cpp/include/fluss.hpp (hunk 5 above)
Comment (charlesdong1991) [reply bindings/cpp/include/fluss.hpp]: changed!

File: bindings/cpp/include/fluss.hpp (hunk 3 above)
Comment (fresh-borzoni) [reply bindings/cpp/include/fluss.hpp:315]: Sure, it works

Comment (charlesdong1991) [reply bindings/cpp/include/fluss.hpp:315]: Created 2 issues

**Review by fresh-borzoni — APPROVED** [review]
@charlesdong1991 Ty for the great work! LGTM 👍

**Review by luoyuxia — APPROVED** [review]
